config/sample-idex.cfg

```diff
@@ -94,3 +100,70 @@ gcode:
     ACTIVATE_EXTRUDER EXTRUDER=extruder1
     SET_DUAL_CARRIAGE CARRIAGE=1
     SET_GCODE_OFFSET Y=15
+
+[gcode_macro calibrate_separation]
+# YOU MUST CALIBRATE YOUR ROTATION_DISTANCE BEFORE THIS IS USEFUL!!!
+# Configure this to draw 2 line segments perpendicular to the dual_carraige axis
+# that meet in the middle of the bed. Both lines should have the same coordinate
+# on the axis of the dual_carraige. If they don't print inline, adjust your endstop
+# settings.
+gcode:
+    G28
+    G90
+    M83
+    T0 ; test T0
+    G1 X120 Y150 Z.2 F4800
+    G1 Y75 E10
+    T1 ; test T1
+    G1 X120 Y0 Z.2
+    G1 Y75 E10
+
+# The following is for if your carraiges have their own part cooling fans.
+# Since this macro overrides M106, you cannot also have a [fan] section defined.
+# If your printer only has one part cooling fan, don't copy this section
+[fan_generic part_fan_0]
+pin: PE5
+
+[fan_generic part_fan_1]
+pin: PD13
+
+[gcode_macro M106]
+description: M106 with support for multiple fans
+gcode:
+    ; if the fanspeed isn't provided it defaults to max speed
+    {% if params.S is defined %}
+        {% set marlinSpeed = params.S|int %}
+    {% else %}
+        {% set marlinSpeed = 255 %}
+    {% endif %}
+
+    ; the fanspeed is put within range or converterd to klipper's speed value
+    {% if marlinSpeed < 0 %}
+        {% set klipperFanSpeed = 0 %}
+    {% elif marlinSpeed > 255 %}
+        {% set klipperFanSpeed = 1 %}
+    {% else %}
+        {% set klipperFanSpeed = marlinSpeed / 255 %}
+    {% endif %}
+
+    ; if P is not a number, it gets set to -1
+    {% if params.P is defined %}
+        {% set target = params.P|int %}
+    {% else %}
+        {% set P = -1 %}
+    {% endif %}
+
+    ; if the target is not a valid fan then all are set
+    {% if target == 0 %}
+            SET_FAN_SPEED FAN=part_fan_0 SPEED={klipperFanSpeed}
+    {% elif target == 1 %}
+            SET_FAN_SPEED FAN=part_fan_1 SPEED={klipperFanSpeed}
+    {% else %}
+            SET_FAN_SPEED FAN=part_fan_0 SPEED={klipperFanSpeed}
+            SET_FAN_SPEED FAN=part_fan_1 SPEED={klipperFanSpeed}
+    {% endif %}
+
+[gcode_macro M107]
+description: Turn off all fans
+gcode:
+    M106 S0
```

Comment:
The fan macros I have. It does a few things differently:
* I follow the `_extruder`/`_extuder1` convention started in the provided sample T0/T1 macros
* I save the fan speed so that you can swop the active extruder's fan when you do a T0/T1
* I set the active extruders fan only by default (This coincides with the auto-park functionality in the provided T0/T1 sections)
* I always disable all fans if `M106 S0` is passed.
* I extended M107 to take an optional `Pn` parameter, to comply with the Marlin usage of it.
* I tried to explain that if you need existing tool fan compatibility, you should use a sacrificial `[fan]` section.

These macros are tested working on my system.

You can change it as you wish

```suggestion
#######################################################################
# Part cooling fans
#######################################################################

# The following is for if your carraiges have their own part cooling fans.
# If your printer only has one part cooling fan, don't copy this section

# Sacrificial "main" fan
# It is switching on an unused pin
# This is so that fan speed is reported correctly on the display and klipper-api(moonraker) interfaces
# If you don't care about that, you can remove this and the "rename_existing" sections in M106 & M107
[fan]
pin: rpi:gpio20

[fan_generic fan_extruder]
pin: PC6

[fan_generic fan_extruder1]
pin: yz:PC6

[gcode_macro M106]
# Only rename_existing if you have a sacrificial [fan] section
rename_existing: M106.1
variable_speed: 0
gcode:
    {% set s = [[params.S|default(255)|int, 255]|min, 0]|max %}
    {% set p = params.P|default(-1)|int %}
    {% set speed = s / 255 %}

    # Save speed variable so you can use it in other macros
    # Access via {printer["gcode_macro M106"].speed}
    SET_GCODE_VARIABLE MACRO=M106 VARIABLE=speed VALUE={speed}

    {% if p == -1 %}
      # Set current active extruder fan
      {% if speed == 0 %}
        # Always turn off al fans if S0 is specified without a specific fan
        SET_FAN_SPEED FAN=fan_extruder SPEED=0
        SET_FAN_SPEED FAN=fan_extruder1 SPEED=0
      {% else %}
        SET_FAN_SPEED FAN=fan_{printer.toolhead.extruder} SPEED={speed}
      {% endif %}
    {% else %}
      # Set specified active extruder fan
      {% if p == 0 %}
        SET_FAN_SPEED FAN=fan_extruder SPEED={speed}
      {% else %}
        SET_FAN_SPEED FAN=fan_extruder1 SPEED={speed}
      {% endif %}     
    {% endif %}

    # Update core Klipper's fan speed
    # Only do this if you have a sacrificial [fan] section
    M106.1 S{s}

[gcode_macro M107]
rename_existing: M107.1
gcode:
    {% set p = params.P|default(-1)|int %}
    M106 S0 P{p}
```

Reply:
OK I like the idea of a sacrificial fan section and I think we can use the T0 and T1 macros to update this value to the active extruder when that changes as well

Reply:
Here is the T0/T1 sections I have that swop the fans between active extruders:
```ini
#######################################################################
# Tool changing
#######################################################################
[gcode_macro _PARK_extruder]
gcode:
    SAVE_GCODE_STATE NAME=park0
    G90
    G1 X283 F10800
    SET_FAN_SPEED FAN=fan_extruder SPEED=0
    RESTORE_GCODE_STATE NAME=park0

[gcode_macro T0]
gcode:
    {% if printer.toolhead.extruder != 'extruder' %}
      SET_FAN_SPEED FAN=fan_extruder SPEED={printer["gcode_macro M106"].speed}
      _PARK_{printer.toolhead.extruder}
      ACTIVATE_EXTRUDER EXTRUDER=extruder
      SET_DUAL_CARRIAGE CARRIAGE=0
      #SET_GCODE_OFFSET Y=0
    {% endif %}

[gcode_macro _PARK_extruder1]
gcode:
    SAVE_GCODE_STATE NAME=park1
    G90
    G1 X-63 F10800
    SET_FAN_SPEED FAN=fan_extruder1 SPEED=0
    RESTORE_GCODE_STATE NAME=park1

[gcode_macro T1]
gcode:
    {% if printer.toolhead.extruder != 'extruder1' %}
      SET_FAN_SPEED FAN=fan_extruder1 SPEED={printer["gcode_macro M106"].speed}
      _PARK_{printer.toolhead.extruder}
      ACTIVATE_EXTRUDER EXTRUDER=extruder1
      SET_DUAL_CARRIAGE CARRIAGE=1
      #SET_GCODE_OFFSET Y=15
    {% endif %}
```

Reply:
Interesting. I am leary of setting the fan speed on tool change. I was thinking the custom m106 remains the only thing that modifies the actual fan speeds and the tool change macro simply updates the sacrificial fan to match the speed of the currently selected extruder.

Reply:
I think, these macros should only do it if a Pn was not specified. 
I also have no idea how this will work with the duplication/mirror modes, we should probably not touch the fans in that case.

Possibly need a gcode_variable that states whether fanspeed should be swapped (and controlled from M106).
I think we could reset `{printer["gcode_macro M106"].speed}` to `-1` (and default to that) to indicate that no fan swapping should occur.
Should then rename it to swap_speed or something for clarity.

Reply:
>  the tool change macro simply updates the sacrificial fan to match the speed of the currently selected extruder.
I missed this. It's a great idea!

:thinking: Now how do I merge both those ideas in a clean way?

Reply:
I updated the tool changing & M106 macros to work as I summarized.
It will now only swap the tool fans if the last call was a `M106 S[>0]` and no P was specified.
Else it will update the sacrifical fan speed to the speed of the current toolhead for informational purposes.

```ini
#######################################################################
# Tool changing
#######################################################################
[gcode_macro _PARK_extruder]
gcode:
    SAVE_GCODE_STATE NAME=park0
    {% if printer["gcode_macro M106"].swap_speed != -1 %}
      SET_FAN_SPEED FAN=fan_extruder SPEED=0
    {% endif %}
    G90
    G1 X283 F10800
    RESTORE_GCODE_STATE NAME=park0

[gcode_macro T0]
gcode:
    # Only do anything if the current tool is not T0
    {% if printer.toolhead.extruder != 'extruder' %}
      {% set fan_speed = printer["gcode_macro M106"].swap_speed %}
      {% if fan_speed != -1 %}
        SET_FAN_SPEED FAN=fan_extruder SPEED={fan_speed}
      {% else %}
        # Update core Klipper's fan speed to the fan speed of the active toolhead
        # Only do this if you have a sacrificial [fan] section
        M106.1 S{printer["fan_generic fan_extruder"].speed * 255}
      {% endif %}
      _PARK_{printer.toolhead.extruder}
      ACTIVATE_EXTRUDER EXTRUDER=extruder
      SET_DUAL_CARRIAGE CARRIAGE=0
      #SET_GCODE_OFFSET Y=0
    {% endif %}

[gcode_macro _PARK_extruder1]
gcode:
    SAVE_GCODE_STATE NAME=park1
    {% if printer["gcode_macro M106"].swap_speed != -1 %}
      SET_FAN_SPEED FAN=fan_extruder1 SPEED=0
    {% endif %}
    G90
    G1 X-63 F10800
    RESTORE_GCODE_STATE NAME=park1

[gcode_macro T1]
gcode:
    # Only do anything if the current tool is not T1
    {% if printer.toolhead.extruder != 'extruder1' %}
      {% set fan_speed = printer["gcode_macro M106"].swap_speed %}
      {% if fan_speed != -1 %}
        SET_FAN_SPEED FAN=fan_extruder1 SPEED={fan_speed}
      {% else %}
        # Update core Klipper's fan speed to the fan speed of the active toolhead
        # Only do this if you have a sacrificial [fan] section
        M106.1 S{printer["fan_generic fan_extruder1"].speed * 255}
      {% endif %}
      _PARK_{printer.toolhead.extruder}
      ACTIVATE_EXTRUDER EXTRUDER=extruder1
      SET_DUAL_CARRIAGE CARRIAGE=1
      #SET_GCODE_OFFSET Y=15
    {% endif %}


#######################################################################
# Part cooling fans
#######################################################################

# The following is for if your carraiges have their own part cooling fans.
# If your printer only has one part cooling fan, don't copy this section

# Sacrificial "main" fan
# It is switching on an unused pin
# This is so that fan speed is reported correctly on the display and klipper-api(moonraker) interfaces
# If you don't care about that, you can remove this and the "rename_existing" sections in M106 & M107
[fan]
pin: rpi:gpio20

[fan_generic fan_extruder]
pin: PC6

[fan_generic fan_extruder1]
pin: yz:PC6

[gcode_macro M106]
# Only rename_existing if you have a sacrificial [fan] section
rename_existing: M106.1
# The variable that controls fan speed swopping if not specifying P parameter
# -1 means the control is disabled, a value of 0-1 is the requested fan speed.
# Access via {printer["gcode_macro M106"].swap_speed}
variable_swap_speed: -1
gcode:
    {% set s = [[params.S|default(255)|int, 255]|min, 0]|max %}
    {% set p = params.P|default(-1)|int %}
    {% set speed = s / 255 %}

    # Set speed to -1 by default
    SET_GCODE_VARIABLE MACRO=M106 VARIABLE=swap_speed VALUE=-1

    {% if p == -1 %}
      # Set current active extruder fan
      {% if speed == 0 %}
        # Always turn off al fans if S0 is specified without a specific fan
        SET_FAN_SPEED FAN=fan_extruder SPEED=0
        SET_FAN_SPEED FAN=fan_extruder1 SPEED=0
      {% else %}
        # Opt into fan speed swop control
        SET_GCODE_VARIABLE MACRO=M106 VARIABLE=swap_speed VALUE={speed}
        SET_FAN_SPEED FAN=fan_{printer.toolhead.extruder} SPEED={speed}
      {% endif %}
    {% else %}
      # Set specified active extruder fan
      {% if p == 0 %}
        SET_FAN_SPEED FAN=fan_extruder SPEED={speed}
      {% else %}
        SET_FAN_SPEED FAN=fan_extruder1 SPEED={speed}
      {% endif %}     
    {% endif %}

    # Update core Klipper's fan speed
    # Only do this if you have a sacrificial [fan] section
    M106.1 S{s}

[gcode_macro M107]
rename_existing: M107.1
gcode:
    {% set p = params.P|default(-1)|int %}
    M106 S0 P{p}
```